src/cascadia/TerminalApp/CommandPalette.cpp

```diff
@@ -1275,15 +1287,42 @@ namespace winrt::TerminalApp::implementation
     void CommandPalette::_updateRecentCommands(const hstring& command)
     {
         const auto recentCommands = ApplicationState::SharedInstance().RecentCommands();
-        // If there aren't and recent commands already in the state, then we
-        // don't need to copy any.
-        const auto countToCopy = std::min(recentCommands ? recentCommands.Size() : 0, CommandLineHistoryLength - 1);
-        std::vector<hstring> newRecentCommands{ countToCopy + 1 };
-        til::at(newRecentCommands, 0) = command;
-        if (countToCopy)
+        // If this is the first time we've opened the commandline mode and
+        // there aren't any recent commands, then just store the new command.
+        if (!recentCommands)
         {
-            recentCommands.GetMany(0, { newRecentCommands.data() + 1, countToCopy });
+            ApplicationState::SharedInstance().RecentCommands(single_threaded_vector(std::move(std::vector{ command })));
+            return;
         }
+
+        const auto numNewRecentCommands = std::min(recentCommands.Size() + 1, CommandLineHistoryLength);
+
+        std::vector<hstring> newRecentCommands;
```

Comment:
On saving I don't think you need to deduplicate everything and have a separate `uniqueCommands`. All you have to do is create the one output vector, add the new command, and then add in commands from `RecentCommands` where they are not equal to the current command.

Reply:
In one of the previous comments @lhecker  suggested to remove deduplication when reading :smile:
I personally prefer to prefer to preserve deduplication in both cases, as IMHO it is a safer programming, usually providing better future compatibility.

Reply:
So, I'm not saying to completely remove deduplication, just that you only need to deduplicate based on the current command. Concretely the following code
```
std::vector<hstring> deduped{};
deduped.reserve(...);
deduped.emplace_back(command);

for (const auto& r : recentCommands){
    if (r != command) { deduped.emplace_back(r); }
}

```

Reply:
Removed the deduplication from loading commands, preserving it here.

Reply:
I prefered to leave the full deduplication for:

1. Backward compatibility (for those in preview who already have duplicate commands)
2. Case where there are several terminals writing the same command

Both are corner cases of course, and yet..

Reply:
It's _theoretically_ for the existing commands list to contain duplicates, but this can only occur if the program was misbehaving and then all bets are off, wouldn't they?
I'm fine with the full deduplication, but technically @Rosefield's code is just as fine IMHO.

Reply:
It is unfortunately difficult to convey things in text. I think that this is great overall, I just thought I saw a case where things could be simplified.

Reply:
Yeah.. conveying things in text sucks! :blush: I try to use emoji - but it is still challenging 🤷 

I absolutely agree that @Rosefield's code is great as well! It is more a matter of taste: some "extra safety" at a price of some in memory checks. Given that this code is invoked only once upon user interaction which is slow by definition, I preferred to pay this price.